*** sccp_pbx.c.bakdd	Mon Nov  6 11:36:37 2006
--- sccp_pbx.c	Mon Nov  6 11:37:45 2006
*************** dial:
*** 785,790 ****
--- 785,792 ----
  	sccp_copy_string(chan->exten, c->dialedNumber, sizeof(chan->exten));
  	sccp_copy_string(d->lastNumber, c->dialedNumber, sizeof(d->lastNumber));
  	sccp_channel_set_calledparty(c, c->dialedNumber, c->dialedNumber);
+ 	/* The 7961 seems to need the dialing callstate to record its directories information. */
+ 	sccp_indicate_nolock(c, SCCP_CHANNELSTATE_DIALING);
  	/* proceed call state is needed to display the called number.
  	The phone will not display callinfo in offhook state */
  	sccp_channel_set_callstate(c, SKINNY_CALLSTATE_PROCEED);
