--- sccp_pbx_orig.c     2006-05-16 10:01:10.000000000 -0400
+++ sccp_pbx.c  2006-05-16 10:02:02.000000000 -0400
@@ -290,6 +290,12 @@
 static int sccp_pbx_answer(struct ast_channel *ast) {
        sccp_channel_t * c = CS_AST_CHANNEL_PVT(ast);
 
+       // if channel type is undefined, set to SCCP
+       if (!ast->type) {
+               sccp_log(1)(VERBOSE_PREFIX_3 "SCCP: Channel type undefined, setting to type 'SCCP'\n");
+               ast->type = "SCCP";
+       }
+
        if (!c || !c->device || !c->line) {
                ast_log(LOG_ERROR, "SCCP: Answered %s but no SCCP channel\n", ast->name);
                return -1; 
