Index: src/pbx_impl/ast/astTrunk.c
===================================================================
--- src/pbx_impl/ast/astTrunk.c	(Revision 4369)
+++ src/pbx_impl/ast/astTrunk.c	(Arbeitskopie)
@@ -31,6 +31,7 @@
 #endif
 struct ast_sched_context *sched = 0;
 struct io_context *io = 0;
+struct ast_format slinFormat = {AST_FORMAT_SLINEAR, {{0},0} };
 
 static PBX_CHANNEL_TYPE *sccp_wrapper_asterisk111_request(const char *type, struct ast_format_cap *format, const PBX_CHANNEL_TYPE * requestor, const char *dest, int *cause);
 static int sccp_wrapper_asterisk111_call(PBX_CHANNEL_TYPE * chan, const char *addr, int timeout);
@@ -430,10 +431,11 @@
 	}
 	//sccp_log((DEBUGCAT_CORE))(VERBOSE_PREFIX_3 "%s: read format: ast->fdno: %d, frametype: %d, %s(%d)\n", DEV_ID_LOG(c->device), ast_channel_fdno(ast), frame->frametype, pbx_getformatname(frame->subclass), frame->subclass);
 	if (frame->frametype == AST_FRAME_VOICE) {
-
-		if (!(ast_format_cap_iscompatible(ast_channel_nativeformats(ast), &frame->subclass.format))) {
-			sccp_wrapper_asterisk111_setReadFormat(c, c->rtp.audio.readFormat);
+#ifdef CS_SCCP_CONFERENCE
+		if (c->conference && (!ast_format_is_slinear(ast_channel_readformat(ast))) ) {
+			ast_set_read_format(ast, &slinFormat);
 		}
+#endif
 	}
 	//
 	//      c = sccp_channel_release(c);
