--- asterisk-1.6.2.16.1/main/rtp.c	2010-10-02 01:01:31.000000000 +0200
+++ asterisk-1.6.2.16.1/main/rtp.c.dd	2011-03-15 02:48:29.703551068 +0100
@@ -1908,6 +1908,8 @@
 		rtp->f.samples = ast_codec_get_samples(&rtp->f);
 		if (rtp->f.subclass == AST_FORMAT_SLINEAR) 
 			ast_frame_byteswap_be(&rtp->f);
+		if (rtp->f.subclass == AST_FORMAT_SLINEAR16) 
+			ast_frame_byteswap_be(&rtp->f);
 		calc_rxstamp(&rtp->f.delivery, rtp, timestamp, mark);
 		/* Add timing data to let ast_generic_bridge() put the frame into a jitterbuf */
 		ast_set_flag(&rtp->f, AST_FRFLAG_HAS_TIMING_INFO);
@@ -1960,6 +1962,7 @@
 	{{1, AST_FORMAT_G726}, "audio", "G726-32", 8000},
 	{{1, AST_FORMAT_ADPCM}, "audio", "DVI4", 8000},
 	{{1, AST_FORMAT_SLINEAR}, "audio", "L16", 8000},
+	{{1, AST_FORMAT_SLINEAR16}, "audio", "L16", 16000},
 	{{1, AST_FORMAT_LPC10}, "audio", "LPC", 8000},
 	{{1, AST_FORMAT_G729A}, "audio", "G729", 8000},
 	{{1, AST_FORMAT_G729A}, "audio", "G729A", 8000},
@@ -2016,6 +2019,7 @@
 	[17] = {1, AST_FORMAT_ADPCM}, /* 22.050 kHz */
 	[18] = {1, AST_FORMAT_G729A},
 	[19] = {0, AST_RTP_CN},		/* Also used for CN */
+	[25] = {1, AST_FORMAT_SLINEAR16}, /* 1 channel, wideband for cisco (-DD) */
 	[26] = {1, AST_FORMAT_JPEG},
 	[31] = {1, AST_FORMAT_H261},
 	[34] = {1, AST_FORMAT_H263},
